File: code/richediting/src/main/java/org/fbme/ide/richediting/adapters/fbnetwork/actions/CollapseAction.java
package org.fbme.ide.richediting.adapters.fbnetwork.actions;

import jetbrains.mps.openapi.editor.cells.EditorCell;
import org.fbme.ide.richediting.adapters.fbnetwork.FunctionBlockController;
import org.fbme.ide.richediting.viewmodel.FunctionBlockView;
import org.fbme.ide.richediting.viewmodel.NetworkComponentView;

import java.util.Set;

public class CollapseAction extends ExpandOrCollapseAction {
    public CollapseAction(EditorCell cell) {
        super(cell.getParent());
    }

    @Override
    public void apply() {
        collapse(selectedFBs);
    }

    private void collapse(Set<NetworkComponentView> selectedComponents) {
        for (NetworkComponentView selectedComponent : selectedComponents) {
            if (selectedComponent instanceof FunctionBlockView) {
                collapse((FunctionBlockView) selectedComponent);
            }
        }
    }

    private void collapse(FunctionBlockView component) {
        FunctionBlockController componentController = (FunctionBlockController) componentsFacility.getController(component);

        preparing(componentController.getFbCell(), component, componentController);
        componentController.expandBlock(false);
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/adapters/fbnetwork/actions/CollapseAction.kt
package org.fbme.ide.richediting.adapters.fbnetwork.actions

import jetbrains.mps.openapi.editor.cells.EditorCell
import org.fbme.ide.richediting.adapters.fbnetwork.FunctionBlockController
import org.fbme.ide.richediting.viewmodel.FunctionBlockView

class CollapseAction(cell: EditorCell) : ExpandOrCollapseAction(cell.parent) {
    fun apply() {
        collapse(selectedFBs.filterIsInstance<FunctionBlockView>())
    }

    private fun collapse(selectedComponents: List<FunctionBlockView>) {
        for (selectedComponent in selectedComponents) {
            collapse(selectedComponent)
        }
    }

    private fun collapse(functionBlock: FunctionBlockView) {
        val componentController = componentsFacility.getController(functionBlock) as FunctionBlockController
        val expandedComponentsController = componentController.expandedComponentsController
//        expandedComponentsController.removeAffectedComponents(component)
        expandedComponentsController.removeFB(functionBlock)
//        expandedComponentsController.removeAffectedSections(component)
    }
}
